README


This archive contains the distribution Calendar-Slots,
version 0.14:

  Create calendars with reserved slots

This software is copyright (c) 2012 by Rodrigo de Oliveira.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.14 Wed Aug 29 12:59:43 CEST 2012
  * FIX crossing slots of different type
  * NEW merge option materialize now treats dates and weekdays the same
  * NEW week_of parameter
  * NEW as_table - for pretty printing

0.13 Mon Aug 27 18:59:10 CEST 2012
  * Fixed slot data missing

0.12 Mon Aug 27 18:46:28 CEST 2012
  * Fixed an overlapping bug, see t/overlap.t for test case

0.11  Mon Aug 20 22:00:23 CEST 2012
  * Added support for $slot->data attribute

0.10  Mon Aug 20 20:05:00 CEST 2012
  * Fixed slot name bug - spaces and other chars removed inadvertently
  * Added weekday to date conversions for slot finding

0.05  Thu Nov 12 09:49:06 CET 2009
  * Fixed failed tests with missing dependency II

0.04  Sun, 08 Nov 2009 10:34:04 +0100
  * Fixed failed tests with missing dependency

0.03  Wed, 01 Oct 2009 09:00:04 +0100
  * Span ranges over dates: start_date, end_date

0.02  Wed, 30 Sep 2009 01:45:04 +0100
  * Fix dependencies

0.01  Sun, 27 Jun 2009 20:59:11 +0100
  * Initial release



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Create calendars with reserved slots",
  "AUTHOR" => "Rodrigo de Oliveira <rodrigolive\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Calendar-Slots",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Calendar::Slots",
  "PREREQ_PM" => {
    "Carp" => 0,
    "DateTime::Format::Strptime" => 0,
    "Moose" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::AttributeHelpers" => 0,
    "Test::More" => "0.9",
    "namespace::autoclean" => 0
  },
  "VERSION" => "0.14",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Calendar/Slots/Slot.pm
package Calendar::Slots::Slot;
{
  $Calendar::Slots::Slot::VERSION = '0.14';
}
use Moose;
use Carp;
use Moose::Util::TypeConstraints;
use namespace::autoclean;

has name     => ( is => 'rw', isa => 'Str' );  # the slot given name
has when     => ( is => 'rw', isa => 'Int', required=>1, );  # weekday num or date
has type     => ( is => 'rw', isa => enum([qw/weekday date/]), required=>1 );  # type of slot
has start    => ( is => 'rw', isa => 'Int' );  # start time
has end      => ( is => 'rw', isa => 'Int' );  # end time
has data     => ( is => 'rw', isa => 'Any' );  # free data for your own use
has _weekday => ( is => 'rw', isa => 'Num' );  # cache

use Calendar::Slots::Utils;

around BUILDARGS => sub {
	my $orig = shift;
	my $class = shift;
	my %args = @_ == 1 && ref $_[0] ? %{ $_[0] || {} } : @_;
	%args    = format_args( %args );

	defined $args{date} and check_date( $args{date} );
	defined $args{weekday} and check_weekday( $args{weekday} );
	defined $args{start} and check_time( $args{start} );
	defined $args{end} and check_time( $args{end} );

	$args{type} = $args{date} ? 'date' : $args{weekday} ? 'weekday' : undef;

	$args{type} ||= length($args{when}) == 8 ? 'date' : 'weekday';
    $args{when} = $args{when}
      || ( $args{type} eq 'date' ? $args{date} : $args{weekday} );
	delete $args{date};
	delete $args{weekday};
	$class->$orig(%args);
};


sub BUILD {
	my $self = shift;
    $self->start > $self->end
      and confess 'Invalid slot: start time is after the end time';
}

sub contains {
    my $self  = shift;
    my %args  = format_args(@_);

	defined $args{date} and check_date( $args{date} );
	defined $args{weekday} and check_weekday( $args{weekday} );
	defined $args{'time'} and check_time( $args{'time'} );

    my $type  = $args{type} || ( $args{date} ? 'date' : 'weekday' );
    my $when  = $args{when}
      || ( $type eq 'date' ? $args{date} : $args{weekday} );
    my $time  = $args{'time'};
    my $start = $args{start};
    my $end   = $args{end};

    $time
      and ( $start or $end )
      and croak 'Parameters start/end and time are mutually exclusive';
    $when or croak 'Missing parameter when';

    if ( $type eq $self->type ) {
        return if $when ne $self->when;
    }
    elsif ( $type eq 'date' && $self->type eq 'weekday' ) {
        $when = parse_dt( '%Y%m%d', $when )->wday;
        $when == 0 and $when = 7;
        return if $when ne $self->when;
    }
    elsif( $when ne $self->weekday ) {
        return ;
    }

    if ($time) {
        return $time >= $self->start && $time < $self->end;
    }
    else {
        return $start > $self->start && $end < $self->end;
    }
    return 1;
}

sub same_weekday {
	my $self = shift;
	my $slot = shift;
	return $self->weekday eq $slot->weekday; 
}

sub weekday {
	my $self = shift;
	my $day;
	if( $self->type eq 'date' ) {
        my $wk = $self->_weekday;
        return $wk if defined $wk;
		my $dt = DateTime->new( $self->ymd_hash );
		return $self->_weekday( $dt->strftime('%u') ); # cache
	}
	else {
		return $self->when;
	}
}

sub same_type {
    my $self = shift;
	my $slot = shift;
	return $self->type eq $slot->type;	
}

sub same_day {
    my $self = shift;
	my $slot = shift;
	return $self->when eq $slot->when;	
}

sub ymd_hash {
    my $self = shift;
	my $when = $self->when;
    return (
        year  => substr( $when, 0, 4 ),
        month => substr( $when, 4, 2 ),
        day   => substr( $when, 6, 2 )
    );
}

sub reschedule {
    my $self = shift;
	my %args = @_;
	if( $self->type eq 'weekday' ) {
		if( my $days = $args{days} ) {
			my $weekday = $self->when + $days;
			$weekday = $weekday > 7 ? $weekday - 7 : $weekday;
			$self->when( $weekday );
		}
	} 
	else {
		my $dt = DateTime->new( $self->ymd_hash );
		$dt->add( %args );
		( my $when = $dt->ymd ) =~ s{/|\-}{}g;
		$self->when( $when );
	}
}

sub numeric {
    my $self = shift;
	if( $self->type eq 'date' ) {
		sprintf("%01d%08d%04d%04d", $self->weekday, 0, $self->start, $self->end );
	} else {
		sprintf("%01d%08d%04d%04d", $self->when, 0, $self->start, $self->end );
	}
}

1;
__END__

=pod

=head1 NAME

Calendar::Slots::Slot - the time-slot object

=head1 VERSION

version 0.14

=head1 SYNOPSIS

	use Calendar::Slots::Slot;
	my $slot = new Calendar::Slots::Slot( date=>'2009-10-22', start=>'20:30', end=>'22:30', name=>'birthday' ); 
	print
		$slot->contains( date=>'2009-10-22', time=>'21:00' )
		? 'I'm busy'
		: 'I'm free then';


=head1 DESCRIPTION

This is the basic class defining a calendar slot. 

=head1 ATTRIBUTES

    has name    => ( is => 'rw', isa => 'Str' );
    has data    => ( is => 'rw', isa => 'Any' );
    has when    => ( is => 'rw', isa => 'Int', required=>1, );
    has start   => ( is => 'rw', isa => 'Int' );
    has end     => ( is => 'rw', isa => 'Int' );
    has type    => ( is => 'rw', isa => 'Str', required=>1 );

=head1 METHODS

=head2 contains( { date=>'YYYY-MM-DD' | weekday=>1..7 }, time=>'HH:MM' )

Returns true or false if the parameters match this time slot.

=head2 numeric

Returns a numeric rendition of the date and time parts, good for sorting.

=head2 weekday

Returns a weekday (1 to 7, Monday to Sunday). Works on both date or weekday 
slots.

=head2 reschedule( add=>Int )

Adds or subtract days to a slot. 

=head2 same_day

Compare two slots and return true if the day is the same. 

=head2 same_weekday( $slot )

Compare two slots and return true if the weekday is the same.

=head2 same_type

Compare two slots and return true if types match.

=head2 ymd_hash

Returns a hash suitable to feed to L<DateTime>:

	DateTime->new( $slot->ymd_hash );

=head1 AUTHOR

Rodrigo de Oliveira C<rodrigolive@gmail.com>

=head1 LICENSE

This library is free software. You can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Calendar/Slots/Utils.pm
package Calendar::Slots::Utils;
{
  $Calendar::Slots::Utils::VERSION = '0.14';
}
use strict;
use warnings;
use DateTime::Format::Strptime;
use Carp;

use vars qw(@ISA @EXPORT);
require Exporter;

@ISA=qw(Exporter);
@EXPORT = qw/format_args check_date check_weekday check_time parse_dt/; 

sub format_args {
    my %args = @_;
    $args{$_} =~ s{[\-|\:|\s]}{}g for grep !/name|data/, keys %args;
    return %args;
}

sub check_time {
    my $time = shift;
    confess 'Invalid time. Must be in the range 00:00 to 24:00'
      unless defined $time and $time >= 0 and $time <= 2400;
}

sub check_weekday {
    my $weekday = shift;
    confess 'Invalid weekday. Must be from 1 (Monday) to 7 (Sunday)'
      unless $weekday >= 1 and $weekday <= 7;
}

sub check_date {
    my $date = shift;
    confess 'Invalid date. Date format must be YYYY-MM-DD'
      unless length($date) == 8;
}

sub parse_dt {
    my ( $format, $date ) = @_;
    use DateTime::Format::Strptime;
    my $parser = DateTime::Format::Strptime->new( pattern => $format );
    return $parser->parse_datetime( $date );
}


1;
__END__

=pod

=head1 NAME

Calendar::Slots::Utils - Calendar::Slots internal machinery

=head1 VERSION

version 0.14

=head1 SYNOPSIS

Nothing here to look at. You may want to take a look at L<Calendar::Slot>.

=head1 METHODS

=head2 check_date

=head2 check_weekday

=head2 check_time

=head2 format_args

=head1 AUTHOR

Rodrigo de Oliveira C<rodrigolive@gmail.com>

=head1 LICENSE

This library is free software. You can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Calendar/Slots.pm
package Calendar::Slots;
{
  $Calendar::Slots::VERSION = '0.14';
}
use Moose;
use MooseX::AttributeHelpers;
use Carp;
use Calendar::Slots::Slot;
use Calendar::Slots::Utils;

has 'overlapping'  => ( is => 'rw', isa => 'Bool', default=>0 );
has 'slots' => (
    metaclass => 'Collection::Array',
    is        => 'ro',
    isa       => 'ArrayRef[Calendar::Slots::Slot]',
    default   => sub { [] },
    provides  => {
        'push'    => 'add_slots',
        'pop'     => 'remove_last_slot',
        'shift'   => 'remove_first_slot',
        'unshift' => 'insert_slots',
        'get'     => 'get_slot_at',
        'set'     => 'set_slot_at',
        'count'   => 'num_slots',
        'empty'   => 'has_slots',
        'clear'   => 'clear_slots',
    }
);

sub slot {
    my $self = shift;
    my @ret;
	for my $slot ( $self->_create_slots( @_ ) ) {
		$self->_validate($slot) unless $self->overlapping;
		my @slots = $self->_merge( {}, $slot, $self->all );
		$self->clear_slots;
		$self->add_slots( @slots );
        push @ret, @slots;
	}
    return @ret;
}

sub _create_slots {
    my $self = shift;

	scalar(@_) == 1 and return $_[0];
    my %args = format_args(@_);

	$args{start_date} and !$args{end_date} and croak 'Missing end_date';
	$args{end_date} and !$args{start_date} and croak 'Missing start_date';

    if ( $args{start_date} && $args{end_date} ) {

        my $start_dt = parse_dt( '%Y%m%d', $args{start_date} )
          or croak "Could not parse start_date '$args{start_date}'";
        my $end_dt = parse_dt( '%Y%m%d', $args{end_date} )
          or croak "Could not parse end_date '$args{end_date}'";
		delete $args{start_date};
		delete $args{end_date};
		my @slots;
		for( my $dt=$start_dt; $dt <= $end_dt; $dt->add( days=>1 ) ) {
			push @slots, $self->_create_slots(date=>$dt->ymd, %args);
		}
		return @slots;	
	}
    elsif ( $args{start} > $args{end} ) {
        my $current = Calendar::Slots::Slot->new( %args, end   => '2400' );
        my $next    = Calendar::Slots::Slot->new( %args, start => '0000' );
		$next->reschedule( days=>1 );
		return ($current, $next);
    }
    else {
        return Calendar::Slots::Slot->new(%args);
    }
}

#check if different slots overlap
sub _validate {
	my $self = shift;
	my $slot = shift;
}

#merge slots that are next to each other
sub _merge {
    my $self = shift;
    my $opts = shift;
    my $new  = shift;
    my @slots = @_;
    unless( scalar @slots ) {
        return $new;
    }
    my $slot = shift @slots;
    if( $opts->{materialize} ) {
        if( ! $slot->same_weekday( $new ) ) {
            return ( $slot, $self->_merge( $opts, $new, @slots ) );
        }
    }
    elsif( !( $slot->same_type($new) && $slot->same_day($new) ) ) {
        # skip this slot
        return ( $slot, $self->_merge( $opts, $new, @slots ) );
    }
    my ( $s1, $s2, $n1, $n2 ) = ( 
        $slot->start, $slot->end, $new->start, $new->end
    );
    #warn join ';', 'new', $new->name, ';', $n1, $n2, '***', $slot->name, $s1, $s2;
    if ( $slot->name eq $new->name ) {
        # s: 10-12, n: 09-12 => merge start
        if ( $n1 < $s1 and $n2 <= $s2 and $n2 >= $s1 ) {
            $slot->start( $new->start );
            return $self->_merge( $opts, $slot, @slots );
        }
        # s: 10-12, n: 11-13 => merge end
        elsif ( $n1 <= $s2 and $n1 >= $s1 and $n2 > $s2 ) {
            $slot->end( $new->end );
            return $self->_merge( $opts, $slot, @slots );
        }
        # s: 10-12, n: 11-12 => discard new
        elsif ( $n1 >= $s1 and $n2 <= $s2 ) {
            return ($slot, @slots);
        }
        # s: 10-12, n: 09-13 => merge all
        elsif ( $n1 < $s1 and $s2 < $n2 ) {
            $slot->start( $new->start );
            $slot->end( $new->end );
            return $self->_merge( $opts, $slot, @slots );
        }
        # s: 10-12, n: 01-05 => add
        else {
            return ($slot, $self->_merge( $opts, $new, @slots) );
        }
    } elsif( ! $self->overlapping ) {
        if ( $slot->start == $new->start and $slot->end == $new->end ) {
            return $self->_merge($opts, $new, @slots);
        }
        elsif ( $n1 <= $s1 and ( $n2 >= $s1 and $n2 <= $s2 ) ) {
            $slot->start( $new->end );
            return ($slot, $self->_merge( $opts, $new, @slots ) );
        }
        elsif ( ( $n1 >= $s1 and $n1 < $s2 ) and $n2 >= $s2 ) {
            $slot->end( $new->start );
            return ($slot, $self->_merge( $opts, $new, @slots ) );
        }
        elsif ( $slot->start < $new->start and $new->end < $slot->end ) {
            my $third = new Calendar::Slots::Slot(
                name  => $slot->name,
                data  => $slot->data,
                when   => $slot->when,
                start => $new->end,
                end   => $slot->end
            );
            $slot->end( $new->start );
            return ( $slot, $third, $new, @slots );
        }
        elsif ( $new->start < $slot->start and $slot->end < $new->end ) {
            return $self->_merge( $opts, $new, @slots );
        }
        else {
            return ($slot, $self->_merge( $opts, $new, @slots) );
        }
    }
    else {
        # overlapping 
        return ($slot, $self->_merge( $opts, $new, @slots) );
    }
}


sub all {
	my $self = shift;
	@{ $self->slots };
}

sub sorted {
	my $self = shift;
	sort {
		$a->numeric <=> $b->numeric
	} $self->all;
}

sub clone {
    my $self = shift;
    my $new = __PACKAGE__->new( %$self, slots=>[] );
    $new->clear_slots;
    $new->slot( %$_ ) for $self->all;
    return $new;
}

sub week_of {
    my ($self, $date ) = @_;
    $date =~ s/\D//g;

    # clone 
    $self = $self->clone;

    # find a monday
    my $dt = parse_dt( '%Y%m%d', $date );
    my $wk = $dt->wday;
    my $ep = $dt->epoch;
    # sunday
    my $sunday_ep = $ep + ( (7-$wk) * 86400 );
    my $sunday = substr DateTime->from_epoch( epoch=>$sunday_ep ), 0, 10;
    # monday
    $wk = 7 if $wk == 0;
    my $monday_ep = $ep - ( ( $wk - 1 ) * 86400 );
    my $monday = substr DateTime->from_epoch( epoch=>$monday_ep ), 0, 10;
    $sunday =~ s/\D//g;
    $monday =~ s/\D//g;
    #  die "$monday - $sunday";
    sub _dump { require YAML; YAML::Dump( \@_ ) };
    return $self->materialize( $monday, $sunday );
}

sub materialize {
    my ($self, $start, $end) = @_;

    # get rid of dates outside this date range
    my @s_wk ;
    my @s_date;
    for my $slot ( $self->all ) {
        if( $slot->type eq 'weekday' ) {
            push @s_wk, $slot;
        }
        elsif( $start <= $slot->when && $slot->when <= $end ) {
            push @s_date, $slot;
        }
    }

    # merge materialized
    $self->clear_slots;
    my @slots = @s_wk;
    for( @s_date ) {
        @slots = $self->_merge( { materialize => 1 }, $_, @slots );  # put the dates first
    }
    $self->clear_slots;
    $self->add_slots( @slots );
    return $self;
}

sub find {
    my $self = shift;
	my %args = @_;
	for my $slot ( grep { $_->type eq 'date' } $self->all ) {
        return $slot
          if $slot->contains(%args);
	}
	for my $slot ( grep { $_->type eq 'weekday' } $self->all ) {
        return $slot
          if $slot->contains(%args);
	}
}

sub name {
	my $slot;
	return $slot->name if $slot = find( @_ );
}

sub as_table {
    my $self = shift;
    require Data::Format::Pretty::Console;
    return Data::Format::Pretty::Console::format_pretty( 
        [ map { { %$_ } } $self->sorted ],
        { table_column_orders=>[[qw/name start end when type _weekday data/]] }
    );
}

1; 

__END__

=pod

=head1 NAME

Calendar::Slots - Manage time slots

=head1 VERSION

version 0.14

=head1 SYNOPSIS

	use Calendar::Slots;
	my $cal = new Calendar::Slots;
	$cal->slot( date=>'2009-10-11', start=>'10:30', end=>'11:30', name=>'busy' ); 
	my $slot = $cal->find( date=>'2009-10-11', time=>'11:00' );
	print $slot->name;    # 'busy'

=head1 DESCRIPTION

This is a simple module to manage a calendar of very generic time slots. Time slots are anything
with a start and end time on a given date or weekday. Time slots cannot overlap. If a new
time slot overlaps another pre-existing time slot, the calendar will acommodate the slot automatically.

It handles two types of slots: fixed dates, or recurring on weekdays. 
When looking for an event, it will search from most specific (date) to more
generic (recurring). That is, if a slot exist for both a date and a weekday, 
it returns the date slot only. 

The calendar is able to compact itself and generate rows that can be easily
stored in a file or database. 

=head1 LIMITATIONS

Some of it current limitations:

=over

=item * No overlapping of time slots. 

=item * If a time-slot spans over midnight, two slots will be created, one for the
selected date until midnight, and another for the next day from midnight until end-time.

=item * It does not handle timezones.

=item * It does not know of daylight-savings or any other DateTime features.

=back

=head1 METHODS

=head2 slot ( name=>Str, { date=>'YYYY-MM-DD' | weekday=>1..7 | start_date/end_date }, start=>'HH:MM', end=>'HH:MM' )

Add a time slot to the calendar.

If the new time slot overlaps an existing slot with the same C<name>, 
the slots are merged and become a single slot. 

If the new time slot overlaps an existing slot with a different C<name>,
it overwrites the previous slot, splitting it if necessary. 

	my $cal = Calendar::Slots->new;
	
	# reserve that monday slot

	$cal->slot( date=>'2009-11-30', start=>'10:30', end=>'11:00', name=>'doctor appointment' ); 

	# create a time slot for a given date

	$cal->slot( date=>'2009-01-01', start=>'10:30', end=>'24:00' ); 

	# create a recurring time slot over 3 calendar days

	$cal->slot( start_date=>'2009-01-01', end_date=>'2009-02-01', start=>'10:30', end=>'24:00' ); 

=head2 find ( { date=>'YYYY-MM-DD' | weekday=>1..7 }, time=>'HH:MM' )

Returns a L<Calendar::Slots::Slot> object for a given .

	$cal->find( weekday=>1, time=>'11:30' );   # find what's on Monday at 11:30

=head2 name 

Shortcut method to L<find|/find> a slot and return a name. 

=head2 sorted 

Returns a  ARRAY of all slot objects in the calendar.

=head2 materialize ( start_date, end_date ) 

Returns an instance of L<Calendar::Slots> with
date slots converted into weekdays for a given
date range.

    my $new_cal = $cal->materialize( 2012_10_22, 2012_10_28 );

=head2 week_of ( date ) 

Returns a materialized instance of L<Calendar::Slots> with actual 
dates merged for the week that comprises 
the passed C<date>.  

    my $week = $cal->week_of( 2012_10_22 );
    $week->find( weekday=>2, time=>10_30 );  # ...

=head2 all 

Returns an ARRAY of all slot objects in the calendar.

=head2 as_table

Returns a console string as a table for the calendar.

Requires that L<Data::Format::Pretty::Console> be installed.

    print $cal->as_table;

=head1 SEE ALSO

L<DateTime::SpanSet>

=head1 TODO

There are many improvements planned for this module, as this is just 
an ALPHA release that allows me to get somethings done at $work...

=over

=item * Other types of recurrence: first Monday, last Friday of September...

=item * Merge several calendars into one.

=item * Create subclasses of Calendar::Slots::Slot for each slot type. 

=item * Better input formatting based on DateTime objects and the such.

=head1 AUTHOR

Rodrigo de Oliveira C<rodrigolive@gmail.com>

=head1 LICENSE

This library is free software. You can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
